pkg/controller/operatingsystemconfig/actuator.go

```diff
@@ -0,0 +1,151 @@
+	cloudConfig, command, err := oscommonactuator.CloudConfigFromOperatingSystemConfig(ctx, log, a.client, osc, generator.IgnitionGenerator())
+	if err != nil {
+		return nil, nil, nil, nil, nil, nil, fmt.Errorf("could not generate cloud config: %w", err)
+	}
+
+	switch purpose := osc.Spec.Purpose; purpose {
+	case extensionsv1alpha1.OperatingSystemConfigPurposeProvision:
+		if !a.useGardenerNodeAgent {
+			return cloudConfig, command, oscommonactuator.OperatingSystemConfigUnitNames(osc), oscommonactuator.OperatingSystemConfigFilePaths(osc), nil, nil, nil
+		}
+		userData, err := a.handleProvisionOSC(ctx, osc)
+		return []byte(userData), nil, nil, nil, nil, nil, err
+
+	case extensionsv1alpha1.OperatingSystemConfigPurposeReconcile:
+		extensionUnits, extensionFiles, err := a.handleReconcileOSC(osc)
+		return cloudConfig, command, oscommonactuator.OperatingSystemConfigUnitNames(osc), oscommonactuator.OperatingSystemConfigFilePaths(osc), extensionUnits, extensionFiles, err
+
+	default:
+		return nil, nil, nil, nil, nil, nil, fmt.Errorf("unknown purpose: %s", purpose)
+	}
+}
+
+func (a *actuator) Delete(_ context.Context, _ logr.Logger, _ *extensionsv1alpha1.OperatingSystemConfig) error {
+	return nil
+}
+
+func (a *actuator) Migrate(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) error {
+	return a.Delete(ctx, log, osc)
+}
+
+func (a *actuator) ForceDelete(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) error {
+	return a.Delete(ctx, log, osc)
+}
+
+func (a *actuator) Restore(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) ([]byte, *string, []string, []string, []extensionsv1alpha1.Unit, []extensionsv1alpha1.File, error) {
+	return a.Reconcile(ctx, log, osc)
+}
+
+func (a *actuator) handleProvisionOSC(ctx context.Context, osc *extensionsv1alpha1.OperatingSystemConfig) (string, error) {
+	writeFilesToDiskScript, err := operatingsystemconfig.FilesToDiskScript(ctx, a.client, osc.Namespace, osc.Spec.Files)
+	if err != nil {
+		return "", err
+	}
+	writeUnitsToDiskScript := operatingsystemconfig.UnitsToDiskScript(osc.Spec.Units)
+
+	script := `#!/bin/bash
+#Fix mis-configuration of dockerd
+mkdir -p /etc/docker
+echo '{ "storage-driver": "devicemapper" }' > /etc/docker/daemon.json
+sed -i '/Environment=DOCKER_SELINUX=--selinux-enabled=true/s/^/#/g' /run/systemd/system/docker.service
+
+# Change existing worker to use docker registry-mirror
+file="/etc/docker/daemon.json"
+if [ $(jq -r 'has("registry-mirrors")' "${file}") == "false" ]; then
+    contents=$(jq -M '. += {"registry-mirrors": ["https://mirror.gcr.io"]}' ${file})
+    echo "${contents}" > ${file}
+fi
+
+systemctl daemon-reload
+systemctl reload docker
+` + writeFilesToDiskScript + `
+` + writeUnitsToDiskScript + `
+
+`
+
+	for _, unit := range osc.Spec.Units {
+		script += fmt.Sprintf(`systemctl enable '%s' && systemctl restart '%s'
```

Comment:
```suggestion
		script += fmt.Sprintf(`systemctl enable '%s' && systemctl restart --no-block '%s'
```
You could add a `--no-block` as we did for other OS extensions ([ref](https://github.com/gardener/gardener-extension-os-gardenlinux/blob/b1d661d31368826551bf30da3df28115cbe2247b/pkg/controller/operatingsystemconfig/actuator.go#L122)). This would prevent issues that the creation of `gardener` user we introduce with https://github.com/gardener/gardener/pull/9077 could be blocked by an error in `gardener-node-init` systemd unit.